### ISSUE TITLE
Updating atom indices in custom forces / interacting with global parameters in plugins

### ISSUE THREAD

**Issue by jlmaccal** [openmm/openmm#3631]
We've run into a major issue in one of our current projects with a tight deadline. I would appreciate whatever support we can get to resolve this quickly.

We have implemented a Bayesian treatment of residual dipolar couplings using a combination of a custom integrator and a `CustomCompoundBondForce`. We have been using this successfully for quite a while now. The implementation is relatively simple and efficient.

In our latest project we are using OpenMM to help determine NMR assignments based on agreement with experimental data. This involves using MCMC steps to permute the mapping between NMR peaks and residues/atoms. We have run into a bit of a show-stopper in that `CustomCompoundBondForce` and friends don't allow for the atom indices to updated by `updateParametersInContext`. Instead, we have to rebuild the context, which takes hundreds of milliseconds—making things prohibitively slow.

A couple of points:
- It seems that only the reference version actually checks that the indices have changed. I believe the CUDA version just silently ignores the changes. This should probably raise an exception.

- What is the origin of this limitation? Is this just to save copying the atom indices, or is there something more going on that would make things too slow?

- How hard would it be to implement a stand-alone version of `CalcRefereneCustomCompoundBondForce` and `CalcCUDACustomCompoundBondForce` that allow the indices to be updated? I've tried looking at the code, but there is a lot going on, particularly with Lepton, that I haven't worked with before.

Finally, one alternative is for me to implement the RDC force as part of our MELD custom CUDA plugin. The main challenge I have is that I don't know how to interact with global parameters and global parameter derivatives in a plugin. For the RDC force, we have a variable number of global parameters depending on the number of different alignment tensors, which adds another level of complexity.
- How can I get an array of global parameter values on the GPU to pass to my kernel?
- Same for the derivatives?
- How is buffering / reduction handled for the global parameter derivatives?
- How do I handle a variable number of global parameters?

**Comment by peastman** [comment]
> I believe the CUDA version just silently ignores the changes. This should probably raise an exception.

Agreed.

> What is the origin of this limitation?

`updateParametersInContext()` is all about tradeoffs.  We want it to be as fast as possible, and every extra piece of information it can update adds cost.  So generally it updates only the most commonly changed information.  There's no fundamental reason it can't do more, but doing so would make it slower.

Updating the atoms involved in a bond is particularly complicated.  At the start of a simulation, it splits up the atoms into molecules based on bonds.  If you allow the bonds to change, that means the definitions of molecules can change.  Rebuilding molecules is expensive, but worse than that, other classes may have already cached the molecule definitions.  For example, MonteCarloBarostat does that.  We don't currently have any mechanism for notifying them that their cached information is no longer valid.

> How can I get an array of global parameter values on the GPU to pass to my kernel?

Here's an example of some code that does that.  It maintains arrays of the values for its parameters both on host and device.  On each step it checks to see if any has changed, and if so it uploads new values to the device.

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/platforms/common/src/CommonKernels.cpp#L338-L348

> Same for the derivatives?

Do you mean for the purpose of computing them?  BondedUtilities mostly takes care of that for you.  Here's the code CustomBondForce uses to set up the derivative calculation.

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/platforms/common/src/CommonKernels.cpp#L315-L320

> How do I handle a variable number of global parameters?

Do you mean that it changes over the course of the simulation?  That isn't allowed.  Or do you just mean it's user configurable?  The above code handles that case.

**Comment by jlmaccal** [comment]
> Updating the atoms involved in a bond is particularly complicated. At the start of a simulation, it splits up the atoms into molecules based on bonds. If you allow the bonds to change, that means the definitions of molecules can change. Rebuilding molecules is expensive, but worse than that, other classes may have already cached the molecule definitions. For example, MonteCarloBarostat does that. We don't currently have any mechanism for notifying them that their cached information is no longer valid.

That makes sense. In our case, the custom force doesn't need to be accounted for when defining the topology of molecules, but I can see how this would be a problem in the general case.

I've been looking at some of the code you lined to, but there's a lot of moving parts and I'm still having trouble understanding how everything fits together.

If I'm going to implement this in CUDA, I would like to incorporate it into our MELD plugin framework, which allows for groups of restraints to be coupled together, so that they are only active depending on their relative energies compared to other restraints in the group. In that case, I don't think I can use bonded utilities, or at least not easily.

Is there a low-level interface for global parameters and derivatives? I can write a kernel that does what it needs to, but I'm not totally sure how to get the parameters and derivatives in/out of it.

If you have time, I think a zoom / teams call would be really helpful. Can you send me an email?

**Comment by peastman** [comment]
> Is there a low-level interface for global parameters and derivatives?

Definitely, it's not really that complicated.  In fact, the simplest thing is just to search for "energyParameterDerivatives" in CudaBondedUtilities.cpp to see what it's doing.  It doesn't show up in very many places.  Here's the basic idea.

You start by calling `addEnergyParameterDerivative()` on the ComputeContext.  That informs it that you'll be computing derivatives for a given parameter.  It allocates a single buffer in which to accumulate contributions to all parameters.  You can get that buffer by calling `getEnergyParamDerivBuffer()` on it.  In your kernel, you add contributions to it exactly the same way you add contributions to the energy buffer.  If the user requests derivatives, the buffers are automatically summed at the end of the energy evaluation.  See `CudaUpdateStateDataKernel::getEnergyParameterDerivatives()`.

> If you have time, I think a zoom / teams call would be really helpful. Can you send me an email?

Sounds good.

**Comment by jlmaccal** [comment]
I decided to implement this for the Reference platform first, but I'm getting a bit stuck.

Let's say my custom force wants to define a parameter `foo`.

- How do I "register" the parameter with the context? I can see the `getParameter` function of `Context`, but how do I register the parameter in the first place?
- How do I flag this parameter as requiring derivative calculations?
- How do I extract this parameter from the context?
- How do I update the derivative?

I know we talked about this today, but I'm still having a hard time seeing how everything fits together.

**Comment by peastman** [comment]
> How do I "register" the parameter with the context?

Your ForceImpl does it by implementing `getDefaultParameters()`:

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/openmmapi/include/openmm/internal/ForceImpl.h#L99-L103

> How do I flag this parameter as requiring derivative calculations?

We talked about this for the CUDA and OpenCL platforms, but the reference platform is much simpler.  You don't need to flag them at all.  The `ReferencePlatform::PlatformData` contains a field for storing the values of derivatives:

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/platforms/reference/include/ReferencePlatform.h#L75

It contains an entry for every parameter, and sets them all to 0 at the start of the energy evaluation:

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/platforms/reference/src/ReferenceKernels.cpp#L172-L173

Unlike the CUDA and OpenCL platforms, which manage a shared workspace buffer for threads to accumulate contributions, this is just a single number for each parameter.  If your force chooses to compute a derivative, it adds it to the appropriate entry in the map.

> How do I extract this parameter from the context?

How do you mean?  If you're asking how to query the parameter value, it's just `context.getParameter()`.

> How do I update the derivative?

Just add your contribution to the entry in the map.  For example,

https://github.com/openmm/openmm/blob/097c5b7543fc638b882438cbf4bf195384c349f6/platforms/reference/src/ReferenceKernels.cpp#L469-L471
